costmap_cspace/include/costmap_cspace/costmap_3d_layer/footprint.h

```diff
@@ -202,112 +219,116 @@ class Costmap3dLayerFootprint : public Costmap3dLayerBase
           }
         }
       }
     }
-    std::vector<bool> unknown;
-    // Get max
-    for (size_t yaw = 0; yaw < map->info.angle; yaw++)
+  }
+  void generateSpecifiedCSpace(
+      CSpace3DMsg::Ptr map,
+      const nav_msgs::OccupancyGrid::ConstPtr& msg,
+      const size_t yaw)
+  {
+    const int ox =
+        std::lround((msg->info.origin.position.x - map->info.origin.position.x) /
+                    map->info.linear_resolution);
+    const int oy =
+        std::lround((msg->info.origin.position.y - map->info.origin.position.y) /
+                    map->info.linear_resolution);
+    const double resolution_scale = msg->info.resolution / map->info.linear_resolution;
+    if (keep_unknown_)
     {
-      if (keep_unknown_)
-      {
-        unknown.resize(msg->data.size());
-        for (size_t i = 0; i < msg->data.size(); i++)
-        {
-          const int gx = std::lround((i % msg->info.width) * resolution_scale) + ox;
-          const int gy = std::lround((i / msg->info.width) * resolution_scale) + oy;
-          if (static_cast<size_t>(gx) >= map->info.width ||
-              static_cast<size_t>(gy) >= map->info.height)
-            continue;
-          // If the cell is unknown in parent map and also updated map,
-          // the cell is never measured.
-          // Never measured cells should be marked unknown
-          // to be correctly processed in the planner.
-          unknown[i] = msg->data[i] < 0 && map->getCost(gx, gy, yaw) < 0;
-        }
-      }
+      unknown_buf_.resize(msg->data.size());
       for (size_t i = 0; i < msg->data.size(); i++)
       {
         const int gx = std::lround((i % msg->info.width) * resolution_scale) + ox;
         const int gy = std::lround((i / msg->info.width) * resolution_scale) + oy;
         if (static_cast<size_t>(gx) >= map->info.width ||
             static_cast<size_t>(gy) >= map->info.height)
           continue;
-        const int8_t val = msg->data[i];
-        if (val < 0)
-        {
-          continue;
-        }
-        else if (val == 0)
-        {
-          int8_t& m = map->getCost(gx, gy, yaw);
-          if (m < 0)
-            m = 0;
-          continue;
-        }
+        // If the cell is unknown in parent map and also updated map,
+        // the cell is never measured.
+        // Never measured cells should be marked unknown
+        // to be correctly processed in the planner.
+        unknown_buf_[i] = msg->data[i] < 0 && map->getCost(gx, gy, yaw) < 0;
+      }
+    }
+    for (size_t i = 0; i < msg->data.size(); i++)
+    {
+      const int gx = std::lround((i % msg->info.width) * resolution_scale) + ox;
+      const int gy = std::lround((i / msg->info.width) * resolution_scale) + oy;
+      if (static_cast<size_t>(gx) >= map->info.width ||
+          static_cast<size_t>(gy) >= map->info.height)
+        continue;
+      const int8_t val = msg->data[i];
+      if (val < 0)
+      {
+        continue;
+      }
```

Comment:
Pretty minor but we can avoid some operations if we move this block to the beginning of the loop.

Reply:
Done